crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -6151,11 +6158,54 @@ mod tests {
             ",
         )?;
 
-        // For these TODOs we need support for `tuple` types:
         let expected_diagnostics = &[
-            // TODO: Should be `RuntimeError | OSError` --Alex
+            "Revealed type is `RuntimeError | OSError`",
+            "Revealed type is `AttributeError | TypeError`",
+        ];
+
+        assert_file_diagnostics(&db, "src/a.py", expected_diagnostics);
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_dynamic_exceptions() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            from typing_extensions import reveal_type
+
+            def foo(
+                x: type[AttributeError],
+                y: tuple[type[OSError], type[RuntimeError]],
+                z: tuple[type[BaseException], ...]
+            ):
+                try:
+                    w
+                except x as e:
+                    reveal_type(e)
+                except y as f:
+                    reveal_type(f)
+                except z as g:
+                    reveal_type(g)
+            ",
+        )?;
+
+        let expected_diagnostics = &[
+            // TODO: these `__class_getitem__` diagnostics are all false positives:
+            // (`builtins.type` is unique at runtime
+            //  as it can be subscripted even though it has no `__class_getitem__` method)
+            "Cannot subscript object of type `Literal[type]` with no `__class_getitem__` method",
+            "Cannot subscript object of type `Literal[type]` with no `__class_getitem__` method",
+            "Cannot subscript object of type `Literal[type]` with no `__class_getitem__` method",
+            "Cannot subscript object of type `Literal[type]` with no `__class_getitem__` method",
```

Comment:
Cc. @charliermarsh -- this is a little edge case in the logic you added in #13579. I think the special-casing that allows e.g. `type[int]` at runtime is buried deep in the interpreter somewhere. According to all the normal rules, it _shouldn't_ be subscriptable... but it is!

```pycon
Python 3.12.4 (main, Jun 12 2024, 09:54:41) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> type[int]
type[int]
>>> type.__getitem__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'type' has no attribute '__getitem__'. Did you mean: '__getstate__'?
>>> type.__class_getitem__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'type' has no attribute '__class_getitem__'
>>> type.__dict__.keys()
dict_keys(['__new__', '__repr__', '__call__', '__getattribute__', '__setattr__', '__delattr__', '__init__', '__or__', '__ror__', 'mro', '__subclasses__', '__prepare__', '__instancecheck__', '__subclasscheck__', '__dir__', '__sizeof__', '__basicsize__', '__itemsize__', '__flags__', '__weakrefoffset__', '__base__', '__dictoffset__', '__name__', '__qualname__', '__bases__', '__mro__', '__module__', '__abstractmethods__', '__dict__', '__doc__', '__text_signature__', '__annotations__', '__type_params__'])
```

Reply:
Yeah this is literally hardcoded directly into `PyObject_GetItem` as a special case 🫣

Reply:
🙃

I'm working on a PR now to add the special case to our semantic model...

Reply:
https://github.com/astral-sh/ruff/pull/13667#issue-2570969264